src/main/java/io/aiven/kafka/auth/LocklessAivenAclAuthorizer.java

```diff
@@ -0,0 +1,276 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.auth;
+
+import java.io.File;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+
+import io.aiven.kafka.auth.audit.AuditorAPI;
+import io.aiven.kafka.auth.json.AivenAcl;
+import io.aiven.kafka.auth.json.reader.AclJsonReader;
+import io.aiven.kafka.auth.json.reader.JsonReader;
+import io.aiven.kafka.auth.json.reader.JsonReaderException;
+
+import kafka.network.RequestChannel.Session;
+import kafka.security.auth.Acl;
+import kafka.security.auth.Authorizer;
+import kafka.security.auth.Operation;
+import kafka.security.auth.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LocklessAivenAclAuthorizer implements Authorizer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AivenAclAuthorizer.class);
+    private File configFile;
+    private AtomicReference<List<AivenAcl>> aclEntries = new AtomicReference<>();
+    private Map<String, Boolean> verdictCache;
+    private JsonReader<AivenAcl> jsonReader;
+    private AuditorAPI auditor;
+    private boolean logDenials;
+
+    // A `stat` system call is needed to check if the configuration was
+    // updated. Because that is expensive changes to the file are checked only
+    // so often. This is the timestamp of the last time the configuration was
+    // checked for modification.
+    private long lastCheckTimestamp = 0;
+
+    // The modified timestamp is used to infer if the file contents changed.
+    private long lastModifiedTimestamp = 0;
+
+    public LocklessAivenAclAuthorizer() {
+        verdictCache = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Reload ACLs from disk.
+     */
+    public boolean reloadAcls() {
+        final boolean changed = reloadConfigIfNecessary();
+        lastCheckTimestamp = getTimestamp();
+        return changed;
+    }
+
+    @Override
+    public void configure(final java.util.Map<String, ?> configs) {
+        final AivenAclAuthorizerConfig config = new AivenAclAuthorizerConfig(configs);
+
+        auditor = config.getAuditor();
+        logDenials = config.logDenials();
+
+        configFile = config.getConfigFile();
+        jsonReader = new AclJsonReader(configFile.toPath());
+        reloadConfigIfNecessary();
+    }
+
+    private long getTimestamp() {
+        return System.nanoTime() / 1000000;  // nanoTime is monotonic, convert to milliseconds
+    }
+
+    private boolean isTimeToCheckConfig(final long now) {
+        return lastCheckTimestamp + 10000 <= now;
+    }
+
+    /**
+     * If the modified timestamp of the file changed, reload it and populate ACL entries.
+     */
+    private synchronized boolean reloadConfigIfNecessary() {
```

Comment:
Could you plz define in the class: 
`private final Object lock = new Object();`
And after that synchronized against it.
 in  `reloadConfigIfNecessary`
 something like that:
`synchronized (lock) {
   .... 
}`